### PR TITLE
Category import

### PR DESCRIPTION
**Pull request by fefrain**

### CODE REVIEW

File: common/survey/categoryImportSummary.js [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+  const headers = getHeaders(summary)
+  return R.pipe(
+    R.keys,
+    R.find(headerName => {
+      const header = headers[headerName]
+      return getHeaderType(header) === type && getHeaderLevelIndex(header) === levelIndex
+    })
+  )(headers)
+}
+
```
Comment (minotogna) [inline common/survey/categoryImportSummary.js]: there's an abuse of the header word. In really ```column``` should be used. header is just a property of the column

File: common/survey/categoryItem.js [hunk 2]
```diff
@@ -12,31 +12,37 @@ const keys = {
...
-      R.prop(language),
-      R.defaultTo(getCode(item))
-    )(item)
+const getLabel = language => item =>
+  R.pipe(
+    getLabels,
+    R.propOr(getCode(item), language),
+  )(item)
+
+const getDescriptions = SurveyUtils.getProp(props.descriptions)
```
Comment (minotogna) [inline common/survey/categoryItem.js]: should we start moving SurveyUtils to ObjectUtils?

File: common/survey/categoryItem.js [hunk 3]
```diff
@@ -12,31 +12,37 @@ const keys = {
...
 const getLabels = SurveyUtils.getProp(props.labels)
 
-const getLabel = language =>
-  item =>
-    R.pipe(
-      getLabels,
-      R.prop(language),
-      R.defaultTo(getCode(item))
-    )(item)
+const getLabel = language => item =>
```
Comment (minotogna) [inline common/survey/categoryItem.js]: don't we have a generic method ```Object.getLabel```? if not, let's add it?

File: common/survey/categoryItem.js [hunk 4]
```diff
@@ -12,31 +12,37 @@ const keys = {
...
+  R.pipe(
+    getLabels,
+    R.propOr(getCode(item), language),
+  )(item)
+
+const getDescriptions = SurveyUtils.getProp(props.descriptions)
+
+const getDescription = language => item =>
```
Comment (minotogna) [inline common/survey/categoryItem.js]: don't we have a generic method ```Object.getDescription```? if not, let's add it?

File: common/survey/categoryItem.js [hunk 5]
```diff
@@ -52,6 +58,12 @@ module.exports = {
   getCode,
   getLabels,
   getLabel,
+  getDescriptions,
+  getDescription,
+  getExtraProp: prop => R.pipe(
+    SurveyUtils.getProp(keys.extra),
```
Comment (minotogna) [inline common/survey/categoryItem.js]: as above, should we start migrating things into ObjectUtils?

File: server/modules/category/service/categoryImportCSVParser.js [hunk 6]
```diff
@@ -0,0 +1,151 @@
...
+  } catch (error) {
+    reject(error)
+  }
+})
+
+const createImportSummary = async (filePath) => {
+  //clean category levels
+  const headers = await _readHeaders(filePath)
+
+  const levelNames = headers.reduce(
```
Comment (minotogna) [inline server/modules/category/service/categoryImportCSVParser.js]: why don't you generate an object with an index as well? e.g. ```[{name:'region',index:0},{name:'district',index:1}]```

File: server/modules/category/service/categoryImportCSVParser.js [hunk 7]
```diff
@@ -0,0 +1,151 @@
...
+      }
+      return accLevelNames
+    },
+    []
+  )
+
+  const headersInfo = headers.reduce(
+    (acc, header) => {
+      if (header.endsWith(columnSuffixes.code)) {
+        const levelIndex = R.pipe(
```
Comment (minotogna) [inline server/modules/category/service/categoryImportCSVParser.js]: if you generate an object (see above comment) you don't need to iterate again on the headers to get it

File: server/modules/category/service/categoryImportCSVParser.js [hunk 8]
```diff
@@ -0,0 +1,151 @@
...
+          R.filter(R.endsWith(columnSuffixes.code)),
+          R.indexOf(header)
+        )(headers)
+
+        const levelName = header.substr(0, header.length - columnSuffixes.code.length)
```
Comment (minotogna) [inline server/modules/category/service/categoryImportCSVParser.js]: don't you have already the level names generated in the first iterations?

File: server/modules/category/service/categoryImportJob.js [hunk 9]
```diff
@@ -0,0 +1,205 @@
...
+    this.logDebug('importing levels')
+
+    const categoryUuid = CategoryImportJobParams.getCategoryUuid(this.params)
+    const summary = CategoryImportJobParams.getSummary(this.params)
+
+    const surveyId = this.getSurveyId()
+    const user = this.getUser()
+
+    await CategoryManager.deleteLevelsByCategory(user, surveyId, categoryUuid, this.tx)
+    let category = await CategoryManager.fetchCategoryByUuid(surveyId, categoryUuid, true, false, this.tx)
```
Comment (minotogna) [inline server/modules/category/service/categoryImportJob.js]: can the ```CategoryManager.deleteLevelsByCategory``` return the category? you won't need the fetch in that case

File: server/utils/file/csvReader.js [hunk 10]
```diff
@@ -0,0 +1,64 @@
...
+      })()
+    }
+
+    const onData = data => {
+      if (headers) {
+        const wasEmpty = queue.isEmpty()
+        queue.enqueue(data)
+        onTotalChange && onTotalChange(++total)
+
+        if (!started || wasEmpty) {
```
Comment (minotogna) [inline server/utils/file/csvReader.js]: why also wasEmpty? started is not enough?

File: common/survey/categoryItem.js (hunk 3 above)
Comment (fefrain) [reply common/survey/categoryItem.js]: This one defaults to the item code... We can't make it generic.

File: common/survey/categoryItem.js (hunk 4 above)
Comment (fefrain) [reply common/survey/categoryItem.js]: Where else will it be needed?

File: server/modules/category/service/categoryImportJob.js (hunk 9 above)
Comment (fefrain) [reply server/modules/category/service/categoryImportJob.js]: Right

File: server/utils/file/csvReader.js (hunk 10 above)
Comment (fefrain) [reply server/utils/file/csvReader.js]: There can be the possibility that the queue becomes empty while there are still rows to read.

File: webapp/loggedin/surveyViews/categoryEdit/actions.js [hunk 11]
```diff
@@ -67,6 +69,27 @@ export const createCategory = () => async (dispatch, getState) => {
...
+export const uploadCategory = (categoryUuid, file) => async (dispatch, getState) => {
+  const formData = new FormData()
+  formData.append('file', file)
+
+  const config = {
+    headers: {
+      'content-type': 'multipart/form-data'
+    }
+  }
+
```
Comment (sgiaccio) [inline webapp/loggedin/surveyViews/categoryEdit/actions.js]: You don't need to set the content-type header, axios does it himself when using FormData

File: server/modules/category/manager/categoryManager.js [hunk 12]
```diff
@@ -43,13 +43,13 @@ const insertCategory = async (user, surveyId, category, client = db) =>
...
-const insertLevel = async (user, surveyId, categoryUuid, level, client = db) =>
+const insertLevel = async (user, surveyId, level, client = db) =>
   await client.tx(async t => {
-    const levelDb = await CategoryRepository.insertLevel(surveyId, categoryUuid, level, t)
+    const levelDb = await CategoryRepository.insertLevel(surveyId, level, t)
 
     await markSurveyDraft(surveyId, t)
 
-    await ActivityLog.log(user, surveyId, ActivityLog.type.categoryLevelInsert, { categoryUuid, level }, t)
+    await ActivityLog.log(user, surveyId, ActivityLog.type.categoryLevelInsert, { level }, t)
```
Comment (sgiaccio) [inline server/modules/category/manager/categoryManager.js]: Can these three operations be done concurrently? i.e. with a `Promise.all` or

```
const levelDbPromise = CategoryRepository.insertLevel(surveyId, level, t)
const markPromise = markSurveyDraft(surveyId, t)
const logPromise = ActivityLog.log(user, surveyId, ActivityLog.type.categoryLevelInsert, { level }, t)

await markPromise
await logPromise
return await levelDbPromise
```

File: server/modules/category/manager/categoryManager.js [hunk 13]
```diff
@@ -65,6 +65,20 @@ const insertItem = async (user, surveyId, item, client = db) =>
...
+    await CategoryRepository.insertItems(surveyId, items, client)
+
+    await markSurveyDraft(surveyId, t)
+
+    const activities = items.map(item => ({
+        type: ActivityLog.type.categoryItemInsert,
+        params: item
+      })
+    )
+    await ActivityLog.logMany(user, surveyId, activities, client)
```
Comment (sgiaccio) [inline server/modules/category/manager/categoryManager.js]: Same as above, can these operations be performed concurrently?

File: server/modules/category/manager/categoryManager.js [hunk 14]
```diff
@@ -165,6 +179,14 @@ const deleteLevel = async (user, surveyId, levelUuid) =>
     await ActivityLog.log(user, surveyId, ActivityLog.type.categoryLevelDelete, { levelUuid }, t)
   })
 
+const deleteLevelsByCategory = async (user, surveyId, categoryUuid, client = db) =>
+  await client.tx(async t => {
+    await CategoryRepository.deleteLevelsByCategory(surveyId, categoryUuid, t)
+    await markSurveyDraft(surveyId, t)
+
+    await ActivityLog.log(user, surveyId, ActivityLog.type.categoryLevelsDelete, { categoryUuid }, t)
```
Comment (sgiaccio) [inline server/modules/category/manager/categoryManager.js]: Same as above, can these operations be performed concurrently?

File: server/modules/category/manager/categoryManager.js (hunk 12 above)
Comment (fefrain) [reply server/modules/category/manager/categoryManager.js]: I'll do it with a Promise.all

File: server/modules/category/manager/categoryManager.js (hunk 13 above)
Comment (fefrain) [reply server/modules/category/manager/categoryManager.js]: yes, sure

File: common/survey/categoryItem.js [hunk 15]
```diff
@@ -12,31 +12,23 @@ const keys = {
 
 const props = {
   code: 'code',
-  labels: 'labels'
+  descriptions: 'descriptions',
+  extra: 'extra',
+  labels: 'labels',
 }
```
Comment (minotogna) [inline common/survey/categoryItem.js]: are you using the keys labels and descriptions? if yes, then we should use ObjectUtils, what do you think?

File: server/modules/category/manager/categoryManager.js (hunk 12 above)
Comment (minotogna) [reply server/modules/category/manager/categoryManager.js]: do you remember that when transactions fails in Promise.all the exception is not catched? I would avoid promise.all

File: server/modules/category/manager/categoryManager.js (hunk 13 above)
Comment (minotogna) [reply server/modules/category/manager/categoryManager.js]: my comment as before

File: server/modules/category/service/categoryImportJob.js (hunk 9 above)
Comment (minotogna) [reply server/modules/category/service/categoryImportJob.js]: ok right, but it still fetches after delete

Comment (fefrain) [reply server/modules/category/service/categoryImportJob.js]: I'll do it... now ;)

File: common/survey/categoryItem.js (hunk 15 above)
Comment (fefrain) [reply common/survey/categoryItem.js]: yes, I'll remove them and use ObjectUtils instead

File: server/modules/category/service/categoryImportJob.js [hunk 16]
```diff
@@ -0,0 +1,202 @@
...
+              [CategoryItem.props.code]: codeLevel,
+              [ObjectUtils.keysProps.labels]: labels[levelName],
+              [ObjectUtils.keysProps.descriptions]: descriptions[levelName],
+              [CategoryItem.props.extra]: extra,
+            }
+          )
+        }
+      }
+    }
+    this.incrementProcessedItems()
```
Comment (minotogna) [inline server/modules/category/service/categoryImportJob.js]: extra are still not imported. will you fix this in the next branch or this one?

Comment (fefrain) [reply server/modules/category/service/categoryImportJob.js]: in the next branch

**Pull request merged**